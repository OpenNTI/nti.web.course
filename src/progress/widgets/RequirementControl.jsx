import './RequirementControl.scss';
import React from 'react';
import PropTypes from 'prop-types';
import { scoped } from '@nti/lib-locale';
import {Flyout} from '@nti/web-commons';
import cx from 'classnames';

import RequirementControlOption from './RequirementControlOption';

const t = scoped('course.components.RequirementControl', {
	required: 'Required',
	optional: 'Optional',
	default: 'Default'
});

const DEFAULT = t('default');
const REQUIRED = t('required');
const OPTIONAL = t('optional');

export default class RequirementControl extends React.Component {
	static propTypes = {
		record: PropTypes.object.isRequired,
		onChange: PropTypes.func,
		className: PropTypes.string
	}

	state = {}

	attachFlyoutRef = x => this.flyout = x

	componentDidMount () {
		const {record} = this.props;

		const basedOnDefault = record.IsCompletionDefaultState;
		const isRequired = record.CompletionRequired;
		const requiredValue = basedOnDefault ? DEFAULT : isRequired ? REQUIRED : OPTIONAL;
		const defaultValue = record.CompletionDefaultState ? REQUIRED : OPTIONAL;

		this.setState({
			value: requiredValue,
			options: [
				{ label: DEFAULT + ' (' + defaultValue + ')', value: DEFAULT },
				{ label: REQUIRED, value: REQUIRED },
				{ label: OPTIONAL, value: OPTIONAL }
			]});
	}

	renderTrigger () {
		const selectedOption = (this.state.options || []).filter(x => x.value === this.state.value)[0] || {};

		const className = cx('require-control-value', this.props.className);

		return <div className={className}><span>{selectedOption.label}</span><i className="icon-chevron-down"/></div>;
	}

	onChange = (newValue) => {
		const {onChange} = this.props;

		if(this.flyout) {
			this.flyout.dismiss();
		}

		this.setState({value: newValue});

		if(onChange) {
			onChange(newValue);
		}
	}

	renderOption = (option) => {
		const isSelected = option.value === this.state.value;

		return <RequirementControlOption key={option.value} option={option} onChange={this.onChange} isSelected={isSelected}/>;
	}

	render () {
		if(!this.state.options) {
			return null;
		}

		const className = cx('require-control', this.props.className);

		return (
			<Flyout.Triggered
				className={className}
				trigger={this.renderTrigger()}
				ref={this.attachFlyoutRef}
				horizontalAlign={Flyout.ALIGNMENTS.LEFT}
			>
				<div>
					{(this.state.options || []).map(this.renderOption)}
				</div>
			</Flyout.Triggered>
		);
	}
}
