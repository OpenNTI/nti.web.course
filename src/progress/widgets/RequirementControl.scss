@import '~@nti/style-common';

.require-control-value {
	height: 30px;
	position: relative;
	font-weight: 600;
	border-radius: 4px;
	line-height: 26px;
	display: inline-block;
	border: 1px solid rgba(151, 151, 151, 0.31);
	padding: 0 12px;
	cursor: pointer;
	color: var(--secondary-grey);
	white-space: nowrap;

	span {
		font-size: 12px;
	}

	i {
		font-size: 10px;
		margin-left: 6px;
	}
}

.require-control {
	.flyout-inner {
		color: var(--secondary-grey);
		font-size: 14px;
		padding: 18px 18px 18px 0;
		min-width: 200px;
	}
}
