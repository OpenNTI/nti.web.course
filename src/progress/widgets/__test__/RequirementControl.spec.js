import React from 'react';
import renderer from 'react-test-renderer';

import RequirementControl from '../RequirementControl';

/* eslint-env jest */
describe('Progress widgets requirement control test', () => {
	test('Default (required)', async () => {
		const record = {
			IsCompletionDefaultState: true,
			CompletionRequired: false,
			CompletionDefaultState: true
		};

		const cmp = renderer.create(
			<RequirementControl record={record}/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Default (optional)', async () => {
		const record = {
			IsCompletionDefaultState: true,
			CompletionRequired: false,
			CompletionDefaultState: false
		};

		const cmp = renderer.create(
			<RequirementControl record={record}/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Required', async () => {
		const record = {
			IsCompletionDefaultState: false,
			CompletionRequired: true,
			CompletionDefaultState: true
		};

		const cmp = renderer.create(
			<RequirementControl record={record}/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});

	test('Optional', async () => {
		const record = {
			IsCompletionDefaultState: false,
			CompletionRequired: false,
			CompletionDefaultState: true
		};

		const cmp = renderer.create(
			<RequirementControl record={record}/>
		);

		const tree = cmp.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
