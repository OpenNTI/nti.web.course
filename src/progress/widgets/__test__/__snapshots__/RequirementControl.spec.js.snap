// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress widgets requirement control test Default (optional) 1`] = `
<div
  aria-haspopup="dialog"
  className="flyout-trigger require-control-value require-control flyout-closed"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
>
  <span>
    Default (Optional)
  </span>
  <i
    className="icon-chevron-down"
  />
</div>
`;

exports[`Progress widgets requirement control test Default (required) 1`] = `
<div
  aria-haspopup="dialog"
  className="flyout-trigger require-control-value require-control flyout-closed"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
>
  <span>
    Default (Required)
  </span>
  <i
    className="icon-chevron-down"
  />
</div>
`;

exports[`Progress widgets requirement control test Optional 1`] = `
<div
  aria-haspopup="dialog"
  className="flyout-trigger require-control-value require-control flyout-closed"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
>
  <span>
    Optional
  </span>
  <i
    className="icon-chevron-down"
  />
</div>
`;

exports[`Progress widgets requirement control test Required 1`] = `
<div
  aria-haspopup="dialog"
  className="flyout-trigger require-control-value require-control flyout-closed"
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onKeyPress={[Function]}
  tabIndex={0}
>
  <span>
    Required
  </span>
  <i
    className="icon-chevron-down"
  />
</div>
`;
